File: scripts/optimize_assets.py
import os
from PIL import Image

def optimize_image(filepath, size=(192, 192)):
    """
    Resizes an image to the specified size and saves it.

    This function opens an image from the given filepath, resizes it using
    LANCZOS resampling for high quality, and saves it back to the same path
    as a PNG file with optimization enabled.

    Args:
        filepath (str): The path to the image file to be optimized.
        size (tuple): A tuple (width, height) representing the target size.
                      Defaults to (192, 192).

    Returns:
        None: This function prints the result to stdout but does not return a value.
    """
    if not os.path.exists(filepath):
        print(f"File not found: {filepath}")
        return

    try:
        with Image.open(filepath) as img:
            print(f"Original size of {filepath}: {os.path.getsize(filepath) / 1024:.2f} KB")
            print(f"Original dimensions: {img.size}")

            # Resize using LANCZOS for high quality downsampling
            img_resized = img.resize(size, Image.Resampling.LANCZOS)

            # Save properly
            img_resized.save(filepath, "PNG", optimize=True)

            print(f"New size of {filepath}: {os.path.getsize(filepath) / 1024:.2f} KB")
            print(f"New dimensions: {img_resized.size}")

    except Exception as e:
        print(f"Error processing {filepath}: {e}")

if __name__ == "__main__":
    # Optimize favicon.png
    optimize_image("public/favicon.png", size=(192, 192))

    # Check star_style.png (just logging, not resizing unless we decide to)
    if os.path.exists("star_style.png"):
         with Image.open("star_style.png") as img:
            print(f"star_style.png dimensions: {img.size}")
            print(f"star_style.png size: {os.path.getsize('star_style.png') / 1024:.2f} KB")

File: scripts/optimize_assets.js
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

async function optimizeImage(filepath, size = { width: 192, height: 192 }) {
    if (!fs.existsSync(filepath)) {
        console.log(`File not found: ${filepath}`);
        return;
    }

    try {
        const stats = fs.statSync(filepath);
        console.log(`Original size of ${filepath}: ${(stats.size / 1024).toFixed(2)} KB`);
        
        const metadata = await sharp(filepath).metadata();
        console.log(`Original dimensions: (${metadata.width}, ${metadata.height})`);

        const buffer = await sharp(filepath)
            .resize({
                width: size.width,
                height: size.height,
                kernel: sharp.kernel.lanczos3
            })
            .png({ quality: 100, compressionLevel: 9 })
            .toBuffer();

        fs.writeFileSync(filepath, buffer);

        const newStats = fs.statSync(filepath);
        console.log(`New size of ${filepath}: ${(newStats.size / 1024).toFixed(2)} KB`);
        
        const newMetadata = await sharp(filepath).metadata();
        console.log(`New dimensions: (${newMetadata.width}, ${newMetadata.height})`);

    } catch (e) {
        console.error(`Error processing ${filepath}: ${e}`);
    }
}

async function main() {
    await optimizeImage("public/favicon.png", { width: 192, height: 192 });

    if (fs.existsSync("star_style.png")) {
        try {
            const stats = fs.statSync("star_style.png");
            const metadata = await sharp("star_style.png").metadata();
            console.log(`star_style.png dimensions: (${metadata.width}, ${metadata.height})`);
            console.log(`star_style.png size: ${(stats.size / 1024).toFixed(2)} KB`);
        } catch (e) {
            console.error(e);
        }
    }
}

main();
